File: packages/cojson/src/crypto/nativeResult.ts
/**
 * Unified result envelope for the native-decision FFI surface.
 *
 * Three native-decision families cross the napi/wasm boundary as a JSON string:
 * the sync content decision (`contentToSend`), the storage write-plan
 * (`planSessionWrite`) and the five group-key writes. They all now emit ONE wire
 * shape, produced by `crates/cojson-core/src/core/native_result.rs`:
 *
 * ```text
 * success:  {"ok":true,"value":<T>}
 * error:    {"ok":false,"kind":"error","message":"<msg>"}
 * ```
 *
 * `parseNativeResult` normalizes that into a {@link NativeResult}, adding a
 * third, TS-side-only case: `unsupported`, for when the native method is absent
 * on the backend (the string is `undefined`). Rust never emits `unsupported`.
 *
 * The point of the envelope is to make a genuine Rust-side failure (`error`)
 * distinguishable from "this backend doesn't have the method" (`unsupported`) —
 * so callers can LOG real errors before falling back to the TS path instead of
 * silently swallowing them (the old bare `try/catch { fall back }` conflated the
 * two).
 */
export type NativeResult<T> =
  | { ok: true; value: T }
  | { ok: false; kind: "unsupported" }
  | { ok: false; kind: "error"; message: string };

/**
 * Parse a native-decision FFI response string into a {@link NativeResult}.
 *
 * - `undefined` input (the native method is not present on this backend) =>
 *   `{ ok: false, kind: "unsupported" }`.
 * - `JSON.parse` throwing, a `{"ok":false,…}` envelope, or a bare
 *   `{"error":"…"}` payload from Rust => `{ ok: false, kind: "error", message }`.
 * - a `{"ok":true,"value":…}` envelope => `{ ok: true, value }`.
 */
export function parseNativeResult<T>(
  json: string | undefined | null,
): NativeResult<T> {
  if (json === undefined || json === null) {
    return { ok: false, kind: "unsupported" };
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(json);
  } catch (e) {
    return {
      ok: false,
      kind: "error",
      message: `native result was not valid JSON: ${String(e)}`,
    };
  }

  if (typeof parsed === "object" && parsed !== null) {
    const obj = parsed as {
      ok?: unknown;
      value?: unknown;
      message?: unknown;
      error?: unknown;
    };

    if (obj.ok === true) {
      return { ok: true, value: obj.value as T };
    }

    if (obj.ok === false) {
      return {
        ok: false,
        kind: "error",
        message:
          typeof obj.message === "string"
            ? obj.message
            : "unknown native error",
      };
    }

    // Defensive: a well-formed bare `{"error":"…"}` payload from Rust.
    if (typeof obj.error === "string") {
      return { ok: false, kind: "error", message: obj.error };
    }
  }

  return {
    ok: false,
    kind: "error",
    message: "malformed native result envelope",
  };
}

File: crates/cojson-core/src/core/native_result.rs
//! Shared serialization for the unified native-decision result envelope.
//!
//! Every redesigned native-decision FFI function — the sync content decision
//! (`content_to_send`), the storage write-plan (`plan_session_write`) and the
//! five group-key writes — returns a JSON string in ONE shape so the TS side can
//! parse them uniformly (`parseNativeResult` in
//! `packages/cojson/src/crypto/nativeResult.ts`).
//!
//! # Wire shape
//!
//! ```text
//! success:  {"ok":true,"value":<T>}
//! error:    {"ok":false,"kind":"error","message":"<msg>"}
//! ```
//!
//! `<T>` is the feature-specific success payload:
//!   * content decision — `NewContentMessage[]`, or `null` when nothing to send;
//!   * storage write-plan — the `SessionWritePlan` object (always present);
//!   * group-key writes — `{"skipped":true}` or
//!     `{"writes":[{"op":"set"|"delete","field":"…","value":"…"?}]}`.
//!
//! The third TS variant `{ok:false,kind:"unsupported"}` is TS-side only (it means
//! the method is absent on the backend); Rust never emits it. Any `Err(String)`
//! from a redesigned function becomes the `error` envelope so a genuine Rust-side
//! failure is distinguishable from a deliberate TS fallback (never silently
//! swallowed).

use serde::Serialize;

/// Envelope a successful value that serializes via serde as
/// `{"ok":true,"value":<value>}`.
pub fn ok_json<T: Serialize>(value: &T) -> String {
    #[derive(Serialize)]
    struct Ok_<'a, T: Serialize> {
        ok: bool,
        value: &'a T,
    }
    match serde_json::to_string(&Ok_ { ok: true, value }) {
        Ok(s) => s,
        Err(e) => error_json(&e.to_string()),
    }
}

/// Envelope an already-serialized JSON value string as
/// `{"ok":true,"value":<value_json>}`. Used when the success payload is produced
/// as a JSON string directly rather than a serde value — e.g. `content_to_send`
/// serializes its content pieces by hand to preserve session insertion order,
/// which routing through `serde_json::Value` would destroy. `value_json` MUST be
/// valid JSON (it is spliced in verbatim); pass `"null"` for the empty case.
pub fn ok_json_raw(value_json: &str) -> String {
    let mut s = String::with_capacity(value_json.len() + 19);
    s.push_str("{\"ok\":true,\"value\":");
    s.push_str(value_json);
    s.push('}');
    s
}

/// Envelope an error message as `{"ok":false,"kind":"error","message":"<msg>"}`.
pub fn error_json(message: &str) -> String {
    #[derive(Serialize)]
    struct Err_<'a> {
        ok: bool,
        kind: &'static str,
        message: &'a str,
    }
    serde_json::to_string(&Err_ {
        ok: false,
        kind: "error",
        message,
    })
    .unwrap_or_else(|_| {
        // A plain string message cannot realistically fail to serialize; emit a
        // static, still-valid envelope so the boundary always yields parseable
        // JSON.
        "{\"ok\":false,\"kind\":\"error\",\"message\":\"native result serialization error\"}"
            .to_string()
    })
}

/// Wrap a `Result<T, String>` into the envelope JSON — the standard bridge from
/// a fallible native function to the wire shape.
pub fn to_result_json<T: Serialize>(result: Result<T, String>) -> String {
    match result {
        Ok(v) => ok_json(&v),
        Err(m) => error_json(&m),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn ok_json_wraps_value() {
        let out: serde_json::Value = serde_json::from_str(&ok_json(&json!({"a": 1}))).unwrap();
        assert_eq!(out, json!({"ok": true, "value": {"a": 1}}));
    }

    #[test]
    fn ok_json_raw_splices_verbatim() {
        assert_eq!(ok_json_raw("null"), "{\"ok\":true,\"value\":null}");
        let out: serde_json::Value =
            serde_json::from_str(&ok_json_raw("[{\"x\":1}]")).unwrap();
        assert_eq!(out, json!({"ok": true, "value": [{"x": 1}]}));
    }

    #[test]
    fn error_json_shape() {
        let out: serde_json::Value = serde_json::from_str(&error_json("boom")).unwrap();
        assert_eq!(
            out,
            json!({"ok": false, "kind": "error", "message": "boom"})
        );
    }

    #[test]
    fn to_result_json_both_arms() {
        let ok: serde_json::Value =
            serde_json::from_str(&to_result_json::<i32>(Ok(7))).unwrap();
        assert_eq!(ok, json!({"ok": true, "value": 7}));
        let err: serde_json::Value =
            serde_json::from_str(&to_result_json::<i32>(Err("nope".to_string()))).unwrap();
        assert_eq!(err, json!({"ok": false, "kind": "error", "message": "nope"}));
    }
}
